File: bindings/tbdex_uniffi/src/messages/cancel.rs
use crate::errors::{Result, TbdexError};
use futures::executor::block_on;
use std::sync::{Arc, RwLock};
use tbdex::{
    json::{FromJson, ToJson},
    messages::cancel::{Cancel as InnerCancel, CancelData},
};
use web5_uniffi_wrapper::dids::bearer_did::BearerDid;

pub struct Cancel(pub Arc<RwLock<InnerCancel>>);

impl Cancel {
    pub fn create(
        to: String,
        from: String,
        exchange_id: String,
        data: CancelData,
        protocol: Option<String>,
        external_id: Option<String>,
    ) -> Result<Self> {
        let close = InnerCancel::create(&to, &from, &exchange_id, &data, protocol, external_id)?;

        Ok(Self(Arc::new(RwLock::new(close))))
    }

    pub fn sign(&self, bearer_did: Arc<BearerDid>) -> Result<()> {
        let mut inner_close = self.0.write().map_err(TbdexError::from_poison_error)?;
        inner_close.sign(&bearer_did.0.clone())?;
        Ok(())
    }

    pub fn from_json_string(json: &str) -> Result<Self> {
        let inner_close = InnerCancel::from_json_string(json)?;

        Ok(Self(Arc::new(RwLock::new(inner_close))))
    }

    pub fn to_json_string(&self) -> Result<String> {
        let inner_close = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(inner_close.to_json_string()?)
    }

    pub fn get_data(&self) -> Result<InnerCancel> {
        let close = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(close.clone())
    }

    pub fn verify(&self) -> Result<()> {
        let cancel = self.0.read().map_err(TbdexError::from_poison_error)?;

        Ok(block_on(cancel.verify())?)
    }
}

File: bound/typescript/src/messages/cancel.ts
import { MessageMetadata } from ".";
import { BearerDid } from "../dids/bearer-did";
import { tbdexError } from "../errors";
import wasm from "../wasm";

export class Cancel {
  readonly metadata: MessageMetadata;
  readonly data: CancelData;
  signature: string;

  constructor(metadata: MessageMetadata, data: CancelData, signature: string) {
    this.metadata = metadata;
    this.data = data;
    this.signature = signature;
  }

  static fromJSONString = (json: string): Cancel => {
    const obj = JSON.parse(json);
    return new Cancel(obj.metadata, obj.data, obj.signature);
  };

  toJSONString = (): string => {
    return JSON.stringify({
      metadata: this.metadata,
      data: this.data,
      signature: this.signature,
    });
  };

  static create = (
    to: string,
    from: string,
    exchangeId: string,
    data: CancelData,
    protocol?: string,
    externalId?: string
  ): Cancel => {
    try {
      let json = wasm.cancel_create(
        to,
        from,
        exchangeId,
        JSON.stringify(data),
        protocol,
        externalId
      );
      let obj = JSON.parse(json);
      return new Cancel(obj.metadata, obj.data, obj.signature);
    } catch (error) {
      throw tbdexError(error);
    }
  };

  sign = (bearerDid: BearerDid) => {
    try {
      const signature = wasm.cancel_sign(
        this.toJSONString(),
        bearerDid.toWASM()
      );
      this.signature = signature;
    } catch (error) {
      throw tbdexError(error);
    }
  };

  verify = async () => {
    try {
      await wasm.cancel_verify(this.toJSONString());
    } catch (error) {
      throw tbdexError(error);
    }
  };
}

export type CancelData = {
  reason?: string;
};
